client/consensus/babe/src/lib.rs

```diff
@@ -521,14 +531,93 @@ where
 	let (worker_tx, worker_rx) = channel(HANDLE_BUFFER_SIZE);
 
 	let answer_requests =
-		answer_requests(worker_rx, config.0, client, babe_link.epoch_changes.clone());
+		answer_requests(worker_rx, config.0, client.clone(), babe_link.epoch_changes.clone());
+
+	let finality_notifications = client.finality_notification_stream();
+	let clean_weights = aux_storage_cleanup(client, finality_notifications);
+
+	// TODO: PROPOSAL
+	// The inner futures are "never-ending" tasks.
+	// Maybe we should replace `join` with `select` to exit from the worker as
+	// soon as one of its inner futures terminates.
+	// The `join` will keep polling the others even if one of them has finished.
 	Ok(BabeWorker {
-		inner: Box::pin(future::join(inner, answer_requests).map(|_| ())),
+		inner: Box::pin(future::join3(inner, answer_requests, clean_weights).map(|_| ())),
 		slot_notification_sinks,
 		handle: BabeWorkerHandle(worker_tx),
 	})
 }
 
+// Remove obsolete block's weight data by leveraging finality notifications.
+// This includes data for all finalized blocks (excluded the most recent one)
+// and all stale branches.
+async fn aux_storage_cleanup<B: BlockT, C>(
+	client: Arc<C>,
+	mut finality_notifications: FinalityNotifications<B>,
+) where
+	C: HeaderBackend<B> + AuxStore,
```

Comment:
Consider `HeaderMetadata` instead of `HeaderBackend` if possible. It is cached, so should be faster.

Reply:
Done